pkg/deployer/baremetal/config.go

```diff
@@ -0,0 +1,244 @@
+	ProcedureDir string `yaml:"procedureDir"`
+}
+
+type Meta struct {
+	StoreAddr  string `yaml:"storeAddr"`
+	ServerAddr string `yaml:"serverAddr"`
+	BindAddr   string `yaml:"bindAddr"`
+	HTTPAddr   string `yaml:"httpAddr"`
+}
+
+type Etcd struct {
+	Artifact *Artifact `yaml:"artifact"`
+}
+
+type Artifact struct {
+	// Local is the local path of binary(greptime or etcd).
+	Local string `yaml:"local"`
+
+	// Version is the release version of binary(greptime or etcd).
+	// Usually, it points to the version of binary of GitHub release.
+	Version string `yaml:"version"`
+}
+
+func (c *Config) Validate() error {
+	if c.Cluster == nil {
+		return fmt.Errorf("empty cluster config")
+	}
+	if c.Etcd == nil {
+		return fmt.Errorf("empty etcd config")
+	}
+
+	if err := c.Cluster.validate(); err != nil {
+		return err
+	}
+	if err := c.Etcd.validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (cluster *Cluster) validate() error {
+	if cluster.Artifact == nil {
+		return fmt.Errorf("cluster artifact field is nil")
+	}
+	if err := cluster.Artifact.validate(); err != nil {
+		return err
+	}
+
+	if cluster.Frontend == nil {
+		return fmt.Errorf("frontend field is nil")
+	}
+	if err := cluster.Frontend.validate(); err != nil {
+		return err
+	}
+
+	if cluster.Meta == nil {
+		return fmt.Errorf("meta field is nil")
+	}
+	if err := cluster.Meta.validate(); err != nil {
+		return err
+	}
+
+	if cluster.Datanode == nil {
+		return fmt.Errorf("datanode field is nil")
+	}
+	if err := cluster.Datanode.validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (etcd *Etcd) validate() error {
+	if etcd.Artifact == nil {
+		return fmt.Errorf("etcd artifact field is nil")
+	}
+	if err := etcd.Artifact.validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO(zyy17): Add the validation of the options.
+func (frontend *Frontend) validate() error {
+	return nil
+}
+
+func (meta *Meta) validate() error {
+	if meta.HTTPAddr == "" {
+		return fmt.Errorf("empty meta http addr")
+	}
+	if err := checkAddr(meta.HTTPAddr); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (datanode *Datanode) validate() error {
+	if datanode.Replicas <= 0 {
+		return fmt.Errorf("invalid replicas '%d'", datanode.Replicas)
+	}
+
+	if datanode.NodeID < 0 {
+		return fmt.Errorf("invalid nodeID '%d'", datanode.NodeID)
+	}
+
+	if datanode.MySQLAddr == "" {
+		return fmt.Errorf("empty datanode mysql addr")
+	}
+	if err := checkAddr(datanode.MySQLAddr); err != nil {
+		return err
+	}
+
+	if datanode.RPCAddr == "" {
+		return fmt.Errorf("empty datanode rpc addr")
+	}
+	if err := checkAddr(datanode.RPCAddr); err != nil {
+		return err
+	}
+
+	if datanode.HTTPAddr == "" {
+		return fmt.Errorf("empty datanode http addr")
+	}
+	if err := checkAddr(datanode.HTTPAddr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (artifact *Artifact) validate() error {
+	if artifact.Version == "" && artifact.Local == "" {
+		return fmt.Errorf("empty artifact")
+	}
```

Comment:
thought: `if artifact.Version != "" && artifact.Local != ""`, which binary will be use to create the cluster?